src/HomogeneousCalorimeter_geo.cpp

```diff
@@ -288,311 +177,181 @@ std::tuple<Volume, Position> build_module(Detector& desc, xml::Collection_t& plm
+  Transform3D      tr_global = RotationZYX(NEEMC_Prot, 0., 0.) * Translation3D(0., 0., 0.);
+  env_vol.setVisAttributes(desc.visAttributes(plm.attr<std::string>(_Unicode(vis_steel_gap))));
 
   // Place frames and mother volume of modules into the world volume
   //
   if (has_envelope) {
-    env.placeVolume(env_vol, tr_global);           // Place the mother volume for all modules
-    env.placeVolume(ring12_vol, tr_global_Oring);  // Place the outer supporting frame
-    env.placeVolume(Sring12_vol, tr_global_Iring); // Place the inner supporting frame
-
-    // env.placeVolume(cabling_support12_V, tr_global_csfront);  // Place the front cabling frame(Only the holes are
-    // visible) env.placeVolume(cabling_support12_V, tr_global_csback);   // Place the back cabling frame(Only the holes
-    // are visible)
-
-    // for(int i = 0 ; i < 12 ; i++)
-    //   {
-    //     hole_pos = RotationZYX((15. + i * 30.)*degree, 0., 0.) * Translation3D(82.5*mm, 675.*mm, 0.);
-    //     cabling_support12_V.placeVolume(hole_V, hole_pos);
-    //     hole_pos = RotationZYX((15. + i * 30.)*degree, 0., 0.) * Translation3D(-82.5*mm, 675.*mm, 0.);
-    //     cabling_support12_V.placeVolume(hole_V, hole_pos);
-    //   }
+    env.placeVolume(env_vol, tr_global);                          // Place the mother volume for all modules
+    env.placeVolume(ring12_vol, tr_global_Oring);                 // Place the outer supporting frame
+    env_vol.placeVolume(inner_support_vol, tr_global_Iring_elli); // Place the version3 inner supporting frame
   }
 
-  // local placement of modules
-  //
-  int  mid = 0, total_id = 0;
-  auto points = epic::geo::fillRectangles({0., 0.}, modSize.x(), modSize.y(), rmin, rmax, phimin, phimax);
+  //=====================================================================
+  // Placing The Modules
+  // Since the inner and outer porfile is not the circle shape,
+  // which means the modules placements can't be simply done by
+  // fillRectangles functions. I hardcode the additional positions
+  // at backward_PbWO4.xml to fill all the gap between circular
+  // placement and supporting structures.
+  //=====================================================================
 
-  Transform3D add_local = RotationZYX(-15. * degree, 0.0, 0.0) * Translation3D(-14.35 * cm, 61.5 * cm, 0.0);
-  auto        modPV     = env_vol.placeVolume(modVol, add_local);
-  modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);
-
-  // Place the modules in the circle
-  for (auto& p : points) {
-    Transform3D tr_local = RotationZYX(-15. * degree, 0.0, 0.0) * Translation3D(p.x(), p.y(), 0.0);
-    modPV    = (has_envelope ? env_vol.placeVolume(modVol, tr_local) : env.placeVolume(modVol, tr_global * tr_local));
-    total_id = id_begin + mid++;
-    modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);
-  }
+  // Add the modules followd the fillRectangles function
+  //
+  int   mid = 0;
```

Comment:
@wdconinc this number is used for the module id, and I use row and column to replaced it. Could you point me how other sub-detectors return row and column ID? I didn't remove this variable in my last commit since I need to change lots of things related if I do so. Thanks!

Reply:
Instead of `int mid`, you are essentially using `int row, column`, so I think you should modify `add_12surface_disk` to return  `std::tuple<int, int, int>` and use it as
```cpp
  for (xml::Collection_t disk_12surface(plm, _Unicode(disk_12surface)); disk_12surface; ++disk_12surface) {
    auto [sector, row, column] = add_12surface_disk(desc, assembly, disk_12surface, sens, sector_id++);
    addModuleNumbers(sector, row, column);
  }
```
where then both `addModuleNumbers` and its loop would need some changes too.

Reply:
Thanks for answering. I tried to adapt it in my last commit, but I also have some questions about how those sector ID, row and column information pass between these functions.

Reply:
Solved in the last commit.